src/game.rs

```diff
@@ -0,0 +1,329 @@
+use crate::cell::Cell;
+use crate::game_error::GameError;
+use num::traits::Num;
+use std::cmp::PartialOrd;
+
+pub type Position = (usize, usize);
+
+///Represents an instance of the Game of Life, it's generated with an initial state of living
+///cells, and it's state can be internally modified to be the next iteration of the game by just calling `next()`
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Game {
+    grid: Vec<Vec<Cell>>,
+    alive_cells: Vec<Position>,
+}
+
+impl Game {
+    /// Arguments:
+    /// - `cells`, a list of initial living cells.
+    /// - `dimensions`, a (usize, usize) which indicates the size of the grid (width, height)
+    ///
+    /// Returns:
+    /// - An instance of Game with size `dimensions` and all cells in `living_cells` alive.
+    /// - An error if `cells` contains out of bound positions with respect to `dimensions`.
+    pub fn from_cells(dimensions: Position, cells: &Vec<Position>) -> Result<Self, GameError> {
+        let (width, height) = dimensions;
+        let mut grid = vec![vec![Cell::new(); width]; height];
+
+        for (cell_i, cell_j) in cells.iter() {
+            //Return error if the cell position is out of bounds
+            if !Self::in_bounds(*cell_i, *cell_j, width, height) {
+                return Err(GameError::OutOfBoundsGridAccess((*cell_i, *cell_j)));
+            }
+            //Otherwise, give life to the cell and return OK
+            grid[*cell_i][*cell_j].give_life();
+        }
+        let alive_cells = cells.clone();
+
+        Ok(Game { grid, alive_cells })
+    }
+
+    /// Updates the internal state to represent the next step in the game according to the
+    /// following rules:
+    /// - Any live cell with fewer than two live neighbours dies, as if caused by underpopulation.
+    /// - Any live cell with two or three live neighbours lives on to the next generation.
+    /// - Any live cell with more than three live neighbours dies, as if by overpopulation.
+    /// - Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
+    pub fn next(&mut self) {
+        let mut to_live = vec![];
+        let mut to_die = vec![];
+        //Iterate over all cells in the grid counting live_neighbors, cells which need to die or
+        //be given life are pushed to the respective vectors.
+        for (i, row) in self.grid.iter().enumerate() {
+            for (j, cell) in row.iter().enumerate() {
+                let live_neighbors = self.live_neighbors((i, j));
+
+                if cell.is_alive() {
+                    if live_neighbors < 2 || live_neighbors > 3 {
+                        to_die.push((i, j));
+                    }
+                } else {
+                    //cell is dead
+                    if live_neighbors == 3 {
+                        to_live.push((i, j));
+                    }
+                }
+            }
+        }
+        self.kill_list(&to_die);
```

Comment:
Can't you just mutate the cells in the previous loop, instead of creating two new lists and iterating over again?